Mathlib/Order/CompactlyGenerated.lean

```diff
@@ -520,70 +552,82 @@ instance (priority := 100) isAtomistic_of_complementedLattice [ComplementedLatti
         exact le_supₛ ⟨ha.of_isAtom_coe_Iic, a.2⟩, fun _ => And.left⟩⟩
 #align is_atomistic_of_complemented_lattice isAtomistic_of_complementedLattice
 
-/-- See Theorem 6.6, Călugăreanu -/
+/-!
+Now we will prove that a compactly generated modular atomistic lattice is a complemented lattice.
+Most explicitly, every element is the complement of a supremum of indepedendent atoms.
+-/
+
+/-- In an atomic lattice, every element `b` has a complement of the form `Sup s`, where each element
+of `s` is an atom. See also `complemented_lattice_of_Sup_atoms_eq_top`. -/
+theorem exists_setIndependent_isCompl_supₛ_atoms (h : supₛ { a : α | IsAtom a } = ⊤) (b : α) :
+    ∃ s : Set α, CompleteLattice.SetIndependent s ∧ IsCompl b (supₛ s) ∧ ∀ ⦃a⦄, a ∈ s → IsAtom a :=
+  by
+  obtain ⟨s, ⟨s_ind, b_inf_Sup_s, s_atoms⟩, s_max⟩ :=
+    zorn_subset
+      { s : Set α | CompleteLattice.SetIndependent s ∧ Disjoint b (supₛ s) ∧ ∀ a ∈ s, IsAtom a }
+      fun c hc1 hc2 =>
+      ⟨⋃₀ c,
+        ⟨CompleteLattice.independent_unionₛ_of_directed hc2.directedOn fun s hs => (hc1 hs).1, ?_,
+          fun a ⟨s, sc, as⟩ => (hc1 sc).2.2 a as⟩,
+        fun _ => Set.subset_unionₛ_of_mem⟩
+  -- swap
+  -- · rw [supₛ_unionₛ, ← supₛ_image, DirectedOn.disjoint_supₛ_right]
+  --   · rintro _ ⟨s, hs, rfl⟩
+  --     exact (hc1 hs).2.1
+  --   · rw [directedOn_image]
+  --     exact hc2.directed_on.mono fun s t => supₛ_le_supₛ
```

Comment:
Weird error here. It complains that it cannot synthesize the placeholder?? The `swap, rw ...` is what is supposed to prove the placeholder, FYI.

Reply:
Last I checked, `obtain` and `have` can't have placeholders in their bodies. You should be able to do
```lean
obtain ⟨s, ⟨s_ind, b_inf_Sup_s, s_atoms⟩, s_max⟩ := by
  refine zorn_subset ... etc etc ...
  rw ... etc etc ...
```
which is cleaner I think.

Reply:
But there's an example of it working on l. 443 of this same file!

Reply:
I see. It seems to be some peculiarity with `obtain`. It works if you split it into a `have` then an `obtain`:
```
  have := zorn_subset
      { s : Set α | CompleteLattice.SetIndependent s ∧ Disjoint b (supₛ s) ∧ ∀ a ∈ s, IsAtom a }
      fun c hc1 hc2 =>
      ⟨⋃₀ c,
        ⟨CompleteLattice.independent_unionₛ_of_directed hc2.directedOn fun s hs => (hc1 hs).1, ?_,
          fun a ⟨s, sc, as⟩ => (hc1 sc).2.2 a as⟩,
        fun _ => Set.subset_unionₛ_of_mem⟩
  obtain ⟨s, ⟨s_ind, b_inf_Sup_s, s_atoms⟩, s_max⟩ := this
  swap
  · rw [supₛ_unionₛ, ← supₛ_image, DirectedOn.disjoint_supₛ_right]
    · rintro _ ⟨s, hs, rfl⟩
      exact (hc1 hs).2.1
    · rw [directedOn_image]
      exact hc2.directedOn.mono @fun s t => supₛ_le_supₛ
```

Reply:
Left a porting note.